File: sdk/python/stellar_insured_sdk/contracts/claims.py
"""Claims contract wrapper with schema-validated payloads."""

from __future__ import annotations

from typing import Any

from ..client import (
    SorobanClient,
    TxOptions,
    to_address,
    to_int128,
    to_uint32,
    to_bool,
    to_string,
    from_scval,
)
from ..schemas import (
    SubmitClaimParams,
    SubmitEvidenceParams,
    VerifyEvidenceParams,
    FlagClaimParams,
    SetPayoutTokenParams,
    SetPauseStateParams,
    PayoutRecord,
)


class ClaimsContract:
    """Typed wrapper for the Claims Soroban contract."""

    def __init__(self, client: SorobanClient, contract_id: str) -> None:
        self._client = client
        self._contract_id = contract_id

    def initialize(self, admin: str, guardian: str, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id,
            "initialize",
            [to_address(admin), to_address(guardian)],
            opts,
        )

    def set_payout_token(
        self, caller: str, token_address: str, opts: TxOptions
    ) -> None:
        params = SetPayoutTokenParams(caller=caller, token_address=token_address)
        self._client.invoke(
            self._contract_id,
            "set_payout_token",
            [to_address(params.caller), to_address(params.token_address)],
            opts,
        )

    def set_max_retries(self, caller: str, max_retries: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id,
            "set_max_retries",
            [to_address(caller), to_uint32(max_retries)],
            opts,
        )

    def submit_claim(
        self,
        policy_address: str,
        claim_id: int,
        policy_id: int,
        amount: int,
        opts: TxOptions,
    ) -> None:
        params = SubmitClaimParams(
            policy_address=policy_address,
            claim_id=claim_id,
            policy_id=policy_id,
            amount=amount,
        )
        self._client.invoke(
            self._contract_id,
            "submit_claim",
            [
                to_address(params.policy_address),
                to_int128(params.claim_id),
                to_int128(params.policy_id),
                to_int128(params.amount),
            ],
            opts,
        )

    def submit_evidence(
        self,
        claim_id: int,
        ipfs_hash: str,
        sensitive: bool,
        description: str,
        submitter: str,
        opts: TxOptions,
    ) -> None:
        params = SubmitEvidenceParams(
            claim_id=claim_id,
            ipfs_hash=ipfs_hash,
            sensitive=sensitive,
            description=description,
            submitter=submitter,
        )
        self._client.invoke(
            self._contract_id,
            "submit_evidence",
            [
                to_int128(params.claim_id),
                to_string(params.ipfs_hash),
                to_bool(params.sensitive),
                to_string(params.description),
                to_address(params.submitter),
            ],
            opts,
        )

    def get_evidence(self, caller: str, evidence_id: int, opts: TxOptions) -> Any:
        result = self._client.invoke(
            self._contract_id,
            "get_evidence",
            [to_address(caller), to_int128(evidence_id)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def get_claim_evidence_ids(self, claim_id: int, opts: TxOptions) -> list[int]:
        result = self._client.invoke(
            self._contract_id,
            "get_claim_evidence_ids",
            [to_int128(claim_id)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def verify_evidence(
        self,
        caller: str,
        evidence_id: int,
        is_valid: bool,
        notes: str,
        opts: TxOptions,
    ) -> None:
        params = VerifyEvidenceParams(
            caller=caller, evidence_id=evidence_id, is_valid=is_valid, notes=notes
        )
        self._client.invoke(
            self._contract_id,
            "verify_evidence",
            [
                to_address(params.caller),
                to_int128(params.evidence_id),
                to_bool(params.is_valid),
                to_string(params.notes),
            ],
            opts,
        )

    def approve_claim(self, claim_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id, "approve_claim", [to_int128(claim_id)], opts
        )

    def reject_claim(self, claim_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id, "reject_claim", [to_int128(claim_id)], opts
        )

    def settle_claim(self, claim_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id, "settle_claim", [to_int128(claim_id)], opts
        )

    def retry_payout(self, claim_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id, "retry_payout", [to_int128(claim_id)], opts
        )

    def get_payout(self, claim_id: int, opts: TxOptions) -> Any:
        result = self._client.invoke(
            self._contract_id,
            "get_payout",
            [to_int128(claim_id)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def get_payout_status(self, claim_id: int, opts: TxOptions) -> str:
        result = self._client.invoke(
            self._contract_id,
            "get_payout_status",
            [to_int128(claim_id)],
            TxOptions(source=opts.source, timeout=opts.timeout, simulate=True),
        )
        return from_scval(result)

    def flag_claim_for_review(
        self, caller: str, claim_id: int, score_adjustment: int, opts: TxOptions
    ) -> None:
        params = FlagClaimParams(
            caller=caller, claim_id=claim_id, score_adjustment=score_adjustment
        )
        self._client.invoke(
            self._contract_id,
            "flag_claim_for_review",
            [
                to_address(params.caller),
                to_int128(params.claim_id),
                to_uint32(params.score_adjustment),
            ],
            opts,
        )

    def report_fraud(self, caller: str, claim_id: int, opts: TxOptions) -> None:
        self._client.invoke(
            self._contract_id,
            "report_fraud",
            [to_address(caller), to_int128(claim_id)],
            opts,
        )

    def set_pause_state(
        self, caller: str, is_paused: bool, reason: str, opts: TxOptions
    ) -> None:
        params = SetPauseStateParams(
            caller=caller, is_paused=is_paused, reason=reason
        )
        self._client.invoke(
            self._contract_id,
            "set_pause_state",
            [
                to_address(params.caller),
                to_bool(params.is_paused),
                to_string(params.reason),
            ],
            opts,
        )

File: sdk/typescript/src/contracts/claims.ts
import {
  SorobanClient,
  TxOptions,
  toAddress,
  toI128,
  toU64,
  toU32,
  toBool,
  toString,
  fromScVal,
} from "../client";
import {
  InitializeClaimsSchema,
  SubmitClaimSchema,
  SubmitEvidenceSchema,
  VerifyEvidenceSchema,
  FlagClaimSchema,
  SetPayoutTokenSchema,
  SetPauseStateSchema,
  type ClaimRecord,
  type PayoutRecord,
} from "../schemas";

export class ClaimsContract {
  constructor(
    private readonly client: SorobanClient,
    private readonly contractId: string
  ) {}

  async initialize(
    admin: string,
    guardian: string,
    opts: TxOptions
  ): Promise<void> {
    const params = InitializeClaimsSchema.parse({ admin, guardian });
    await this.client.invoke(
      this.contractId,
      "initialize",
      [toAddress(params.admin), toAddress(params.guardian)],
      opts
    );
  }

  async setPayoutToken(
    caller: string,
    tokenAddress: string,
    opts: TxOptions
  ): Promise<void> {
    const params = SetPayoutTokenSchema.parse({
      caller,
      token_address: tokenAddress,
    });
    await this.client.invoke(
      this.contractId,
      "set_payout_token",
      [toAddress(params.caller), toAddress(params.token_address)],
      opts
    );
  }

  async setMaxRetries(
    caller: string,
    max: number,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "set_max_retries",
      [toAddress(caller), toU32(max)],
      opts
    );
  }

  async submitClaim(
    policyAddress: string,
    claimId: bigint,
    policyId: bigint,
    amount: bigint,
    opts: TxOptions
  ): Promise<void> {
    const params = SubmitClaimSchema.parse({
      policy_address: policyAddress,
      claim_id: claimId,
      policy_id: policyId,
      amount,
    });
    await this.client.invoke(
      this.contractId,
      "submit_claim",
      [
        toAddress(params.policy_address),
        toI128(params.claim_id),
        toI128(params.policy_id),
        toI128(params.amount),
      ],
      opts
    );
  }

  async submitEvidence(
    claimId: bigint,
    ipfsHash: string,
    sensitive: boolean,
    description: string,
    submitter: string,
    opts: TxOptions
  ): Promise<void> {
    const params = SubmitEvidenceSchema.parse({
      claim_id: claimId,
      ipfs_hash: ipfsHash,
      sensitive,
      description,
      submitter,
    });
    await this.client.invoke(
      this.contractId,
      "submit_evidence",
      [
        toI128(params.claim_id),
        toString(params.ipfs_hash),
        toBool(params.sensitive),
        toString(params.description),
        toAddress(params.submitter),
      ],
      opts
    );
  }

  async getEvidence(
    caller: string,
    evidenceId: bigint,
    opts: TxOptions
  ): Promise<unknown> {
    const result = await this.client.invoke(
      this.contractId,
      "get_evidence",
      [toAddress(caller), toI128(evidenceId)],
      { ...opts, simulate: true }
    );
    return fromScVal(result);
  }

  async getClaimEvidenceIds(
    claimId: bigint,
    opts: TxOptions
  ): Promise<bigint[]> {
    const result = await this.client.invoke(
      this.contractId,
      "get_claim_evidence_ids",
      [toI128(claimId)],
      { ...opts, simulate: true }
    );
    return fromScVal<bigint[]>(result);
  }

  async verifyEvidence(
    caller: string,
    evidenceId: bigint,
    isValid: boolean,
    notes: string,
    opts: TxOptions
  ): Promise<void> {
    const params = VerifyEvidenceSchema.parse({
      caller,
      evidence_id: evidenceId,
      is_valid: isValid,
      notes,
    });
    await this.client.invoke(
      this.contractId,
      "verify_evidence",
      [
        toAddress(params.caller),
        toI128(params.evidence_id),
        toBool(params.is_valid),
        toString(params.notes),
      ],
      opts
    );
  }

  async approveClaim(
    claimId: bigint,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "approve_claim",
      [toI128(claimId)],
      opts
    );
  }

  async rejectClaim(
    claimId: bigint,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "reject_claim",
      [toI128(claimId)],
      opts
    );
  }

  async settleClaim(
    claimId: bigint,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "settle_claim",
      [toI128(claimId)],
      opts
    );
  }

  async retryPayout(
    claimId: bigint,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "retry_payout",
      [toI128(claimId)],
      opts
    );
  }

  async getPayout(
    claimId: bigint,
    opts: TxOptions
  ): Promise<PayoutRecord> {
    const result = await this.client.invoke(
      this.contractId,
      "get_payout",
      [toI128(claimId)],
      { ...opts, simulate: true }
    );
    return fromScVal<PayoutRecord>(result);
  }

  async getPayoutStatus(
    claimId: bigint,
    opts: TxOptions
  ): Promise<string> {
    const result = await this.client.invoke(
      this.contractId,
      "get_payout_status",
      [toI128(claimId)],
      { ...opts, simulate: true }
    );
    return fromScVal<string>(result);
  }

  async flagClaimForReview(
    caller: string,
    claimId: bigint,
    scoreAdjustment: number,
    opts: TxOptions
  ): Promise<void> {
    const params = FlagClaimSchema.parse({
      caller,
      claim_id: claimId,
      score_adjustment: scoreAdjustment,
    });
    await this.client.invoke(
      this.contractId,
      "flag_claim_for_review",
      [
        toAddress(params.caller),
        toI128(params.claim_id),
        toU32(params.score_adjustment),
      ],
      opts
    );
  }

  async reportFraud(
    caller: string,
    claimId: bigint,
    opts: TxOptions
  ): Promise<void> {
    await this.client.invoke(
      this.contractId,
      "report_fraud",
      [toAddress(caller), toI128(claimId)],
      opts
    );
  }

  async setPauseState(
    caller: string,
    isPaused: boolean,
    reason: string,
    opts: TxOptions
  ): Promise<void> {
    const params = SetPauseStateSchema.parse({
      caller,
      is_paused: isPaused,
      reason,
    });
    await this.client.invoke(
      this.contractId,
      "set_pause_state",
      [toAddress(params.caller), toBool(params.is_paused), toString(params.reason)],
      opts
    );
  }
}
